src/annotator.cpp

```diff
@@ -0,0 +1,361 @@
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::UNITS, model->units(u))));
+        }
+        for (size_t i = 0; i < model->units(u)->unitCount(); ++i) {
+            std::string prefix;
+            std::string reference;
+            double exponent;
+            double multiplier;
+            model->units(u)->unitAttributes(i, reference, prefix, exponent, multiplier, id);
+            if (!id.empty()) {
+                idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::UNIT, std::make_pair(model->units(u), i))));
+            }
+        }
+        if (model->units(u)->isImport() && model->units(u)->importSource() != nullptr) {
+            id = model->units(u)->importSource()->id();
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::IMPORT, model->units(u)->importSource())));
+        }
+    }
+    // Components.
+    for (size_t c = 0; c < model->componentCount(); ++c) {
+        listComponentIdsAndItems(model->component(c), idList);
+    }
+    // Encapsulation.
+    id = model->encapsulationId();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::ENCAPSULATION, std::string())));
+    }
+
+    return idList;
+}
+
+void Annotator::build(const ModelPtr &model)
+{
+    mPimpl->isBuilt = false;
+    removeAllIssues();
+    mPimpl->mIdList.clear();
+    mPimpl->mIdList = listIdsAndItems(model);
+    mPimpl->isBuilt = true;
+    mPimpl->mModel = model;
+}
+
+AnyItem Annotator::item(const std::string &id)
+{
+    AnyItem item;
+    if (!mPimpl->isBuilt) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Please call the Annotator::build function before attempting to access items by their id.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+    if (mPimpl->mIdList.empty()) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("The id map is empty; the supplied model has no id attributes.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+
+    // Retrieve the item from the idList.
+    ItemList::iterator it;
+    it = mPimpl->mIdList.find(id);
+
+    if (it == mPimpl->mIdList.end()) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Could not find an item with id='" + id + "' in the model.");
+        issue->setLevel(libcellml::Issue::Level::WARNING);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        // Adding the requested id string and this issue to the map, so that future requests (ie: to the .second)
+        // will return the same IssuePtr rather than creating a new one.
+        mPimpl->mIdList.insert(std::make_pair(id, item));
```

Comment:
no, really don't like this approach. Essentially adding a new ID to the model in a not very nice way. But also looking at this, it means that this method will always return an `item` and then the user needs to check if that item is of type `ISSUE` to see if it was successful. That seems very wrong to me...but maybe I'm missing something about those any items...

Reply:
No, that's not how it works - there's no ID added to the model at all.  This is used so that we can keep track of the same request for nonexistent IDs, and only return one issue for them.  Otherwise we'd be generating a duplicate issue each time.

Reply:
yeah, sorry, not really added to the model, but if you just look at the id list in the annotator as the list of items in the current model that can be annotated the user is forced to check that each item is at least not an `issue` before being able to use it...again, this is starting to look like the use of `any` is less than optimal.

Reply:
It's nothing to do with the use of `any`? That's only so that we can return a pointer reasonably efficiently (rather than having a list of each different pointer type in the argument list of the function).  It's easy enough to remove!  We could just as easily have an `Annotator::Type::NOT_FOUND` enum or similar, with a `nullptr` as the object.  It would still need the user to check for type though, same as before.  But a side note - is viewing/printing a list of ids something a user would want?  I don't do that at the moment?